Issues to compile dune from scratch

When opening an issue, following issue instructions i began the process to add a test & open PR. I followed the [hacking](https://github.com/ocaml/dune/blob/master/doc/hacking.rst) instruction from an empty switch, but wasn't able to easily compile.

## Expected Behavior

Update guidelines/makefile to ease install from scratch, on an empty *environment*.

## Actual Behavior

### hacking.rst
Following hacking guidelines, i went for the simple (on an empty switch).
```
$ make dev
$ ./dune.exe build @check
$ ./dune.exe runtest
```

`make dev` builds `dune.exe`, but fails because of missing dependencies
```
$ make dev
ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot unix.cma boot/libs.ml boot/duneboot.ml
./.duneboot.exe
./_boot/dune.exe build @install
File "otherlibs/dune-rpc-lwt/src/dune", line 4, characters 38-46:
4 |  (libraries result csexp dune_rpc lwt lwt.unix unix)
                                          ^^^^^^^^
Error: Library "lwt.unix" not found.
-> required by library "dune-rpc-lwt" in
   _build/default/otherlibs/dune-rpc-lwt/src
-> required by _build/default/META.dune-rpc-lwt
-> required by _build/install/default/lib/dune-rpc-lwt/META
-> required by _build/default/dune-rpc-lwt.install
-> required by alias install
make: *** [Makefile:53: dev] Error 1
```
However, it builds  `dune.exe`, but checking that everything compile or launching test (the 2 other commands) fail as there is other missing dependencies (ppx_inline_test.config, ppx_inline_test.config, core_bench.inline_benchmarks, lwt, stdio, etc.).

#### Reproduction
```
$ opam switch create test-sw --empty # or an already empty switch
$ git clone git@github.com:ocaml/dune.git
$ cd dune
$ git reset --hard e8ad4cd5db8b9b2
$ make dev
$ ./dune.exe build @check
$ ./dune.exe runtest
```

### Using `make dev-switch`
After discussion with @emillon, i was advised to use `make dev-switch`. Here it fails too, but for another reason: opam can't install dependencies at switch creation as they depend on dune.
```
test "test-sw" = "/tmp/dune" || \
	opam switch create -y . 4.14.0 --deps-only --with-test

<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><><><>
Switch invariant: ["ocaml-base-compiler" {= "4.14.0"} | "ocaml-system" {= "4.14.0"}]
The following actions will be performed:
=== install 38 packages
  ∗ base                  v0.15.1 [required by ppx_expect]
[...]
#=== ERROR while compiling cppo.1.6.9 =========================================#
# context     2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.0 | https://opam.ocaml.org
# path        /tmp/dune/_opam/.opam-switch/build/cppo.1.6.9
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p cppo -j 6
# exit-code   1
# env-file    ~/.opam/log/cppo-78510-577e06.env
# output-file ~/.opam/log/cppo-78510-577e06.out
### output ###
# bwrap: execvp dune: No such file or directory


<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build cppo                1.6.9
│ λ build csexp               1.5.1
│ λ build jane-street-headers v0.15.0
│ λ build ocaml-compiler-libs v0.12.4
│ λ build pp                  1.1.2
│ λ build ppx_derivers        1.2.1
│ λ build re                  1.10.4
│ λ build result              1.5
│ λ build sexplib0            v0.15.1
│ λ build stdlib-shims        0.3.0
[...]
```
#### Reproduction
```
$ opam switch create test-sw --empty # or an already empty switch
$ git clone git@github.com:ocaml/dune.git
$ cd dune
$ git reset --hard e8ad4cd5db8b9b2
$ make dev-switch
```

### Using `make dev-switch` on an empty local switch
Finally, I decided to create a local empty switch to not let the makefile create one, and go directly to dependencies install. It worked well! I was able to build dune, check everything, and run testsuite. The only remaining issue is a missing binary on a test: `node`
```
$ ./dune.exe build @check
File "test/blackbox-tests/test-cases/dune", line 18, characters 8-19:
18 |   (deps %{bin:node})
             ^^^^^^^^^^^
Error: Program node not found in the tree or in PATH
```
#### Reproduction
```
$ opam switch create test-sw --empty # or an already empty switch
$ git clone git@github.com:ocaml/dune.git
$ cd dune
$ git reset --hard e8ad4cd5db8b9b2
$ opam switch create . --empty
$ make dev-switch
$ make dev
$ ./dune.exe build @check
$ ./dune.exe runtest
```
However, it wasn't a blocker for new test addition.

## Specifications

- Version of `dune` (output of `dune --version`): main branch, e8ad4cd5db8b9b2
- Version of `ocaml` (output of `ocamlc --version`): irrelevant
- Operating system (distribution and version): debian

## Additional information 

Full log can be found [here](https://gist.github.com/rjbou/bb480986eee45dae9e0699513d9c8dd7).

Thanks!

--- a/Makefile
+++ b/Makefile
@@ -6,25 +6,6 @@ DESTDIR_ARG := $(if $(DESTDIR),--destdir $(DESTDIR),)
 INSTALL_ARGS := $(PREFIX_ARG) $(LIBDIR_ARG) $(DESTDIR_ARG)
 BIN := ./_boot/dune.exe
 
-# Dependencies used for testing dune, when developed locally and
-# when tested in CI
-TEST_DEPS := \
-lwt \
-cinaps \
-"csexp>=1.3.0" \
-js_of_ocaml \
-js_of_ocaml-compiler \
-"mdx>=2.3.0" \
-menhir \
-ocamlfind \
-ocamlformat.$$(awk -F = '$$1 == "version" {print $$2}' .ocamlformat) \
-"odoc>=2.4.0" \
-"ppx_expect>=v0.16.0" \
-ppx_inline_test \
-ppxlib \
-ctypes \
-"utop>=2.6.0" \
-"melange>=4.0.0-414"
 # Dependencies recommended for developing dune locally,
 # but not wanted in CI
 DEV_DEPS := \
@@ -68,31 +49,33 @@ uninstall:
 .PHONY: reinstall
 reinstall: uninstall install
 
+.PHONY: install-ocamlformat
 install-ocamlformat:
 	opam install -y ocamlformat.$$(awk -F = '$$1 == "version" {print $$2}' .ocamlformat)
 
-dev-depext:
-	opam depext -y $(TEST_DEPS)
-
 .PHONY: dev-deps
-dev-deps:
-	opam install -y $(TEST_DEPS)
+dev-deps: install-ocamlformat
+	opam install -y . --deps-only --with-dev-setup
 
 .PHONY: coverage-deps
 coverage-deps:
 	opam install -y bisect_ppx
 
 .PHONY: dev-deps-sans-melange
-dev-deps-sans-melange:
-	opam install -y $(TEST_DEPS)
+dev-deps-sans-melange: dev-deps
 
 .PHONY: dev-switch
 dev-switch:
 	opam update
 # Ensuring that either a dev switch already exists or a new one is created
-	test "$(shell opam switch show)" = "$(shell pwd)" || \
-		opam switch create -y . $(TEST_OCAMLVERSION) --deps-only --with-test
-	opam install -y --update-invariant ocaml.$(TEST_OCAMLVERSION) $(TEST_DEPS) $(DEV_DEPS)
+	if test -d _opam ; then \
+		opam install -y --update-invariant ocaml.$(TEST_OCAMLVERSION); \
+	else \
+		opam switch create -y . $(TEST_OCAMLVERSION) --no-install ; \
+	fi
+	opam install -y . --deps-only --with-test --with-dev-setup
+	$(MAKE) install-ocamlformat
+	opam install -y $(DEV_DEPS)
 
 .PHONY: test
 test: $(BIN)

--- a/doc/hacking.rst
+++ b/doc/hacking.rst
@@ -24,6 +24,12 @@ To create a directory-local opam switch with the dependencies necessary to build
 
   $ make dev-switch
 
+This can also be used to keep the switch updated when dependencies change.
+
+The ``Makefile`` also has a ``make dev-deps`` which will install just the
+dependencies used by tests. These are marked ``{ with-dev-setup }`` in Dune's
+opam file.
+
 Bootstrapping
 =============
 

--- a/dune.opam
+++ b/dune.opam
@@ -46,4 +46,19 @@ depends: [
   ("ocaml" {>= "4.08"} | ("ocaml" {>= "4.02" & < "4.08~~"} & "ocamlfind-secondary"))
   "base-unix"
   "base-threads"
+  "lwt" { with-dev-setup & os != "win32" }
+  "cinaps" { with-dev-setup }
+  "csexp" { with-dev-setup & >= "1.3.0" }
+  "js_of_ocaml" { with-dev-setup & os != "win32" }
+  "js_of_ocaml-compiler" { with-dev-setup & os != "win32" }
+  "mdx" { with-dev-setup & >= "2.3.0" & os != "win32" }
+  "menhir" { with-dev-setup & os != "win32" }
+  "ocamlfind" { with-dev-setup & os != "win32" }
+  "odoc" { with-dev-setup & >= "2.4.0" & os != "win32" }
+  "ppx_expect" { with-dev-setup & >= "v0.16.0" & os != "win32" }
+  "ppx_inline_test" { with-dev-setup & os != "win32" }
+  "ppxlib" { with-dev-setup & os != "win32" }
+  "ctypes" { with-dev-setup & os != "win32" }
+  "utop" { with-dev-setup & >= "2.6.0" & os != "win32" }
+  "melange" { with-dev-setup & >= "4.0.0-414" & os != "win32" }
 ]

--- a/dune.opam.template
+++ b/dune.opam.template
@@ -8,4 +8,19 @@ depends: [
   ("ocaml" {>= "4.08"} | ("ocaml" {>= "4.02" & < "4.08~~"} & "ocamlfind-secondary"))
   "base-unix"
   "base-threads"
+  "lwt" { with-dev-setup & os != "win32" }
+  "cinaps" { with-dev-setup }
+  "csexp" { with-dev-setup & >= "1.3.0" }
+  "js_of_ocaml" { with-dev-setup & os != "win32" }
+  "js_of_ocaml-compiler" { with-dev-setup & os != "win32" }
+  "mdx" { with-dev-setup & >= "2.3.0" & os != "win32" }
+  "menhir" { with-dev-setup & os != "win32" }
+  "ocamlfind" { with-dev-setup & os != "win32" }
+  "odoc" { with-dev-setup & >= "2.4.0" & os != "win32" }
+  "ppx_expect" { with-dev-setup & >= "v0.16.0" & os != "win32" }
+  "ppx_inline_test" { with-dev-setup & os != "win32" }
+  "ppxlib" { with-dev-setup & os != "win32" }
+  "ctypes" { with-dev-setup & os != "win32" }
+  "utop" { with-dev-setup & >= "2.6.0" & os != "win32" }
+  "melange" { with-dev-setup & >= "4.0.0-414" & os != "win32" }
 ]

